profile/endless-ca-trust-nss.sh

```diff
@@ -0,0 +1,6 @@
+# Add Endless CA certificate to user's NSS database
+
+if [ -d "$HOME/.pki/nssdb" ]; then
```

Comment:
Just a quick note - I think this only affects the current user - maybe we need a copy of .pki to /etc/skel too?

Reply:
It definitely does only affect the current user. I hadn't considered `/etc/skel`, but that could be nice. FWIW, I doubt `/etc/skel` would be of much use since you've almost certainly created your primary user by the time you get around to having this package installed.

I'm specifically not trying to add the certificate to the OS (where it would help a lot more since you'd have `/etc/skel` populated before the first boot experience) because I don't want every person who installs Endless to trust our internal CA certificate.

Reply:
Yeah that's fair - but maybe it would be good to then add a CLI tool to have the current user enroll the CA if it's a new one and you already have the package installed?

Reply:
Yeah, that's part of this PR. The script is `endless-ca-trust-nss`. The thing you're commenting on above is the "let's try to make that automatic by calling it during login". I.e., if you install `endless-ca-cert`, then you'll get `endless-ca-trust-nss` run on your next login. The test for `~/.pki/nssdb` is to try to avoid doing that for system users that happen to have normal login shells which pull in /etc/profile.

Oh, but now I see that putting an empty NSS DB in /etc/skel would help since only "real" users should be using getting the files from /etc/skel. That would make sure the above conditional worked even if you hadn't run chrome/chromium yet.

Reply:
I looked at creating an empty NSS db in /etc/skel/.pki/nssdb, but I'm worried that it's not OK to simply copy it to another place since it appears the directory is written into a config file in the DB.

Instead, I could make an empty `/etc/skel/.pki`, change the about conditional to just check for `$HOME/.pki` and change `endless-ca-trust-nss` so it automatically creates the DB for you. What do you think?

Reply:
Hmm - maybe we can tell system users apart from human users by the UID without messing with skel? >=1000 I agree that the 99.99% case of this script ever being used is on a dev system with 1 human user besides the auto-created shared (uid 1000), so I don't think we need to get pre-occupied with smoothly/automatically handling multi-user cases.

Ultimately however, you won't get me massively excited by solving this problem because I think we should just stop all of this and move quickly to LE DNS-authorised certs, wildcard certs with a reverse proxy so there's no security mix-up (a compromised "leaf" node with a cert copy impersonating others), HTTP authorised certs with some kind of reverse proxy or firewall hole to get the challenge accessible to LE's hosts, etc.

Reply:
@dbnicholson Yeah that would make sense to me.

@ramcq For internal and high-profile systems, I really think that the default way that LE validates domain ownership is an absolutely bad idea from a security standpoint (your certificate validation should not have root access to your box and have capabilities to change your endpoint webserver configuration). From my point of view, being loose w/ this approach on fringe servers is fine but I have to keep pushing for a more-isolated approach for LE certs in the high-value sections of our infra which we still haven't figured out. With that all said, @dbnicholson is the owner of most of this so he gets the final say but I don't think we should just stick LE on everything and call it a day.

Reply:
@sgnn7 Don't get me wrong - certbot is _insane_. I use dehydrated running as a non-root user, and have statically configured the certification folder, etc on some hosts, on others I have a priveleged host that runs limited privelege per-host ssh hooks to sync and deploy the new certs. And - there are a number of LE verification methods, including opening a socket, placing a file on HTTP, adding zones to DNS, etc, so we just need to think about this and we can absolutely figure out a way to do this without causing a security issue on the host.

(it strikes my that anyone running ansible or a bastion host you run ansible from likely already has these priveleges, so there might be a way to benefit from this to get/generate/provision the certs too)

In any case, I _do_ absolutely think we should stick LE certs on everything (in a sensible/responsible way - not certbot and public HTTP etc) and call it a day. The first rule of being a CA is: don't be a CA. :)

These hosts are predominantly all VPN'd anyway, which entails mutual authorisation between the user and the Endless internal network. So: given this, are we trying to protect against browsers whinging, or a trojan host inside our secure network and a successful compromise of our internal DNS/IP/MAC/etc? I think LE is sufficient for both, given each host must then be addressed/attacked/etc individually. I don't understand the additional value provided by our own boutique CA for 30-40 humans (assuming charitably everyone isn't already totally inured/trained to click through and ignore the SSL errors resulting from our previous self-signed certs), especially considering the additional engineering resources it has, and continues to, consume...

(In any case, I don't know why this script can't just create the .pki folder for a human user rather than having to edit the skel contents for every Endless system and maintain that delta from Debian, etc.)

Reply:
I'm updating the script to create `~/.pki/nssdb` and the profile script will call it when the UID is greater than or equal to 1000. I think that's reasonable and I agree with @ramcq that it's better than screwing around with `/etc/skel`.

I'll leave the discussion about using LE certificates for internal hosts for some other time since the infrastructure for using them on non-public hosts here does not exist but our custom CA does. I'm sure there is a sensible way to do this, but I don't have time to take on another project like that.

Reply:
@dbnicholson Sounds good!

@ramcq We can talk more in #tools about LE :)